src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -496,182 +1249,315 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
-    if (status != imcode::Success)
+    // If there is an ActivePresetHandle set, find the preset in the pending presets list that matches the ActivePresetHandle
+    // attribute. If a preset is not found with the same presetHandle, return INVALID_IN_STATE. If there is no ActivePresetHandle
+    // attribute set, continue with other checks.
+    uint8_t buffer[kPresetHandleSize];
+    MutableByteSpan activePresetHandle(buffer);
+
+    err = delegate->GetActivePresetHandle(activePresetHandle);
+
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(Zcl, "Warning: AbsMaxHeatSetpointLimit missing using default");
+        return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::InvalidInState);
     }
-    status = MinHeatSetpointLimit::Get(endpoint, &MinHeatSetpointLimit);
-    if (status != imcode::Success)
+
+    if (!activePresetHandle.empty())
     {
-        MinHeatSetpointLimit = AbsMinHeatSetpointLimit;
+        uint8_t count = CountPresetsInPendingListWithPresetHandle(delegate, activePresetHandle);
+        if (count == 0)
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::InvalidInState);
+        }
     }
 
-    status = MaxHeatSetpointLimit::Get(endpoint, &MaxHeatSetpointLimit);
-    if (status != imcode::Success)
+    // For each preset in the pending presets list, check that the preset does not violate any spec constraints.
+    for (uint8_t i = 0; true; i++)
     {
-        MaxHeatSetpointLimit = AbsMaxHeatSetpointLimit;
-    }
+        PresetStructWithOwnedMembers pendingPreset;
+        err = delegate->GetPendingPresetAtIndex(i, pendingPreset);
 
-    // Make sure the user imposed limits are within the manufacturer imposed limits
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            break;
+        }
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Zcl,
+                         "emberAfThermostatClusterCommitPresetsSchedulesRequestCallback: GetPendingPresetAtIndex failed with error "
+                         "%" CHIP_ERROR_FORMAT,
+                         err.Format());
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::InvalidInState);
+        }
 
-    // https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3725
-    // Spec does not specify the behavior is the requested setpoint exceeds the limit allowed
-    // This implementation clamps at the limit.
+        bool isPendingPresetWithNullPresetHandle = pendingPreset.GetPresetHandle().IsNull();
+        if (isPendingPresetWithNullPresetHandle)
+        {
+            // If the presetHandle for a preset is null, the device should set a unique presetHandle.
+            const uint8_t handle[] = { static_cast<uint8_t>(pendingPreset.GetPresetScenario()) };
+            pendingPreset.SetPresetHandle(DataModel::MakeNullable(ByteSpan(handle)));
```

Comment:
This doesn't make sense.  In what way is this unique?   It's not.

Reply:
For now we support only one preset of each type. Will file a followup to clean this up. does that work?

Reply:
Filed https://github.com/project-chip/connectedhomeip/issues/34556